File: LeetCode/Medium/1717-Maximum_Score_From_Removing_Substrings/1717-MaximumScoreFromRemovingSubstrings.cpp
#include <algorithm>
#include <stack>
#include <string>

using namespace std;

class Solution
{
public:
    string remove(string s, string sub)
    {
        stack<char> stack;

        for (char c : s)
        {
            if (c == sub[1] && !stack.empty() && stack.top() == sub[0])
                stack.pop();
            else
                stack.push(c);
        }

        string res;

        while (!stack.empty())
        {
            res.push_back(stack.top());
            stack.pop();
        }

        reverse(res.begin(), res.end());
        return res;
    }

    int maximumGain(string s, int x, int y)
    {
        int ans = 0;
        string first;
        string second;
        if (y > x)
        {
            first = "ba";
            second = "ab";
        }
        else
        {
            first = "ab";
            second = "ba";
        }

        string new_s = remove(s, first);
        ans += ((s.size() - new_s.size()) / 2) * max(x, y);
        ans += ((new_s.size() - remove(new_s, second).size()) / 2) * min(x, y);

        return ans;
    }
};

File: LeetCode/Medium/1717-Maximum_Score_From_Removing_Substrings/1717-MaximumScoreFromRemovingSubstrings.py
class Solution:
    def maximumGain(self, s: str, x: int, y: int) -> int:
        def remove(s, sub):
            stack = []

            for c in s:
                if c == sub[1] and stack and stack[-1] == sub[0]:
                    stack.pop()
                else:
                    stack.append(c)

            return "".join(stack)

        ans = 0
        if y > x:
            first = "ba"
            second = "ab"
        else:
            first = "ab"
            second = "ba"

        new_s = remove(s, first)
        ans += ((len(s) - len(new_s)) // 2) * max(x, y)
        ans += ((len(new_s) - len(remove(new_s, second))) // 2) * min(x, y)

        return ans


a = Solution()
print(a.maximumGain(s="cdbcbbaaabab", x=4, y=5))
print(a.maximumGain(s="aabbaaxybbaabb", x=5, y=4))
